File: PA2_Client/src/edu/ucdenver/tournament/Tournament.java
package edu.ucdenver.tournament;

import java.io.*;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.ArrayList;
import java.util.List;

public class Tournament implements Serializable{
    private String name;
    private LocalDate startDate;
    private LocalDate endDate;

    private List<Match> listMatches;
    private List<Team> listTeams;
    private List<Referee> listReferees;
    private List<Country> participatingCountries;

    //--------------------------------------------------
    //                  constructors
    //--------------------------------------------------
    public Tournament(String name, LocalDate startDate, LocalDate endDate){
        this.name = name;
        this.startDate = startDate;
        this.endDate = endDate;

        this.listMatches = new ArrayList<>();
        this.listTeams = new ArrayList<>();
        this.listReferees = new ArrayList<>();
        this.participatingCountries = new ArrayList<>();
    }


    //--------------------------------------------------
    //                    getters
    //--------------------------------------------------

    public List<Match> getListMatches(){
        return listMatches;
    }

    public List<Team> getListTeams(){
        return listTeams;
    }

    public List<Referee> getListReferees(){
        return listReferees;
    }

    public List<Country> getParticipatingCountries(){
        return participatingCountries;
    }

    public String getName() {
        return name;
    }

    //--------------------------------------------------
    //                    setters
    //--------------------------------------------------

    public void setName(String name) {
        this.name = name;
    }

    public LocalDate getStartDate() {
        return this.startDate;
    }

    public void setStartDate(LocalDate startDate) {
        this.startDate = startDate;
    }

    public LocalDate getEndDate() {
        return this.endDate;
    }

    public void setEndDate(LocalDate endDate) {
        this.endDate = endDate;
    }


    //--------------------------------------------------
    //                  class methods
    //--------------------------------------------------

    public static Tournament loadFromFile(String filename){
        ObjectInputStream ois = null;
        Tournament tournament = null;

        try{
            ois = new ObjectInputStream(new FileInputStream(filename));
            tournament = (Tournament) ois.readObject();
        }
        catch (Exception e){
            e.printStackTrace();
            tournament = new Tournament(null, null, null);
        }
        finally{
            if (ois != null){
                try{
                    ois.close();
                }
                catch (IOException ioe){
                    ioe.printStackTrace();
                }
            }
        }
        return tournament;
    }

    public void saveToFile(String filename){
        ObjectOutputStream oos = null;

        try{
            oos = new ObjectOutputStream(new FileOutputStream(filename));
            oos.writeObject(this);
        }
        catch (IOException ioe){
            ioe.printStackTrace();
        }
        finally {
            if (oos != null){
                try{
                    oos.close();
                }
                catch (IOException ioe){
                    ioe.printStackTrace();
                }
            }
        }
    }

    public void addCountry(String countryName) throws IllegalArgumentException{
        //creating a Country object
        Country country = null;

        //checking to see if the country already exists in the tournament
        for(Country c : participatingCountries){
            if(c.getCountryName().equalsIgnoreCase(countryName)){
                throw new IllegalArgumentException("The country '" + countryName + "' already exists.");
            }
        }

        //if the country does not exist, add it
        country = new Country(countryName);
        this.participatingCountries.add(country);
    }

    public void addTeam(String name, String countryName) throws IllegalArgumentException{
        //Creating country and team objects
        Country country = null;
        Team team = null;

        //checking if the team already exists in the tournament
        for(Team t : listTeams){
            if(t.getName().equalsIgnoreCase(name)){
                throw new IllegalArgumentException("The team '" + name + "' already exists.");
            }
        }

        //checking if the country already exists in the tournament
        for (Country c : participatingCountries){
            if(c.getCountryName().equalsIgnoreCase(countryName)){
                country = c;
            }
        }

        //if the country already exists in the tournament, add the new team
        if(country != null) {
            team = new Team(name, country);
            this.listTeams.add(team);
        }
        //if the country does not exist in the tournament, throw an error
        else if(country == null){
            throw new IllegalArgumentException("The country '" + countryName + "' does not exist.");
        }
    }

    public void addReferee(String name, String countryName) throws IllegalArgumentException{
        //creating country and referee objects
        Country country = null;
        Referee referee = null;

        //checking to see if the referee already exists in the tournament
        for (Referee r : listReferees){
            if(r.getName().equalsIgnoreCase(name)){
                throw new IllegalArgumentException("The referee '" + name + "' already exists.");
            }
        }

        //checking to see if the country already exists in the tournament
        for (Country c : participatingCountries){
            if(c.getCountryName().equalsIgnoreCase(countryName)){
                country = c;
            }
        }

        //if the country exists but the referee does not, add the new referee
        if(referee == null && country != null){
            referee = new Referee(name, country);
            this.listReferees.add(referee);
        }

        //if the country does not exist, throw an error
        else if (country == null){
            throw new IllegalArgumentException("The country '" + countryName + "' does not exist.");
        }
    }

    public void addPlayer(String teamName, String playerName, int age, double height, double weight)
            throws IllegalArgumentException{
        //creating a Team object
        Team team = null;
        Player player = null;

        //checking to see if the team already exists in the tournament
        for (Team t : listTeams){
            if(t.getName().equalsIgnoreCase(teamName)){
                team = t;
            }
        }

        //if the team exists, add the player
        if(team != null) {
            //checking to see if the player already exists in the tournament
            for (Player p : team.getSquad()){
                if(p.getName().equalsIgnoreCase(playerName)){
                    throw new IllegalArgumentException("The player '" + playerName + "' already exists.");
                }
            }
            team.addPlayer(playerName, age, height, weight);
        }
        else if(team == null){
            throw new IllegalArgumentException("The team '" + teamName + "' does not exist.");
        }
    }

    public void addMatch(LocalDate dateTime, String teamAName, String teamBName) throws IllegalArgumentException{
        //creating two Team objects, one for team A and one for team B
        Team teamA = null;
        Team teamB = null;

        if(teamAName == teamBName){
            throw new IllegalArgumentException("Team A and Team B cannot be the same.");
        }

        //checking to see if team A and team B exist in the tournament
        for (Team t : listTeams){
            if(t.getName().equalsIgnoreCase(teamAName)){
                teamA = t;
            }
            if(t.getName().equalsIgnoreCase(teamBName)){
                teamB = t;
            }
        }

        //if the teams exist in the tournament, add the match
        if(teamA != null && teamB != null){
            Match match = null;

            //checking to see if the match already exists in the tournament and throwing an error if it does
            for (Match m: listMatches){
                if(m.getDateTime().equals(dateTime)){
                    throw new IllegalArgumentException("This match already exists");
                }
            }

            //adding match if it does not already exist
            match = new Match(dateTime, teamA, teamB);
            listMatches.add(match);
        }
        else if(teamA == null || teamB == null){
            throw new IllegalArgumentException("One of the teams entered does not exist.");
        }
    }

    public void addRefereeToMatch(LocalDate dateTime, String refereeName) throws IllegalArgumentException{
        //creating a Referee and Match object
        Referee referee = null;
        Match match = null;

        //checking to see if the referee exists in the tournament
        for(Referee r : listReferees){
            if(r.getName().equalsIgnoreCase(refereeName)){
                referee = r;
            }
        }

        //checking to see if the match exists in the tournament
        for(Match m : listMatches){
            if(m.getDateTime().equals(dateTime)){
                match = m;
            }
        }

        //if both the referee and the match exist, add the referee to the match
        if(referee != null && match != null){
            if(referee.getCountry().getCountryName().equalsIgnoreCase(match.getTeamA().getTeam().getCountry().getCountryName()) ||
                    referee.getCountry().getCountryName().equalsIgnoreCase(match.getTeamB().getTeam().getCountry().getCountryName())){
                throw new IllegalArgumentException("Could not add referee to match. " +
                        "The referee cannot be from the same country as either team A or team B.");
            }
            match.addReferee(referee);
        }
        //throwing an error if the referee doesnt exist
        else if(referee == null){
            throw new IllegalArgumentException("The referee '" + refereeName + "' does not exist.");
        }
        //throwing an error if the match doesnt exist
        else if(match == null){
            throw new IllegalArgumentException("This match does not exist.");
        }
        //handling other errors
        else{
            throw new IllegalArgumentException("Error!");
        }
    }

    public void addPlayerToMatch(LocalDate dateTime, String teamName, String playerName)
            throws IllegalArgumentException{
        //creating Match, Team, and Player objects
        Match match = null;
        Team team = null;
        Player player = null;

        //checking to see if the match exists in the tournament
        for (Match m : listMatches){
            if(m.getDateTime().equals(dateTime)){
                match = m;
            }
        }

        //checking to see if the team exists in the tournament
        for(Team t : listTeams){
            if(t.getName().equalsIgnoreCase(teamName)){
                team = t;
            }
        }

        //if the team exists in the tournament, search for the player
        if(team != null){
            for (Player p : team.getSquad()){
                if(p.getName().equalsIgnoreCase(playerName)){
                    player = p;
                }
            }
        }

        //if the match, the team, and the player all exist in the tournament, add the player to the match
        if(match != null && team != null && player != null){
            match.addPlayer(player, team);
        }
        //throwing an error if the match doesnt exist
        else if(match == null){
            throw new IllegalArgumentException("This match does not exist.");
        }
        //throwing an error if the team doesnt exist
        else if(team == null){
            throw new IllegalArgumentException("The team '" + teamName + "' does not exist.");
        }
        //throwing an error if the player doesnt exist
        else if(player == null){
            throw new IllegalArgumentException("The player '" + playerName +
                    "' does not exist or is not part of this team.");
        }
        //handling other errors
        else{
            throw new IllegalArgumentException("Error!");
        }
    }

    public void setMatchScore(LocalDate dateTime, int score1, int score2) throws IllegalArgumentException{
        //creating a Match object
        Match match = null;
        LocalDate localDate = LocalDate.now(ZoneId.systemDefault());

        //check to see if the match exists in the tournament
        for(Match m : listMatches){
            if(m.getDateTime().equals(dateTime)){
                match = m;
            }
        }

        //if the match exists, set the score
        if(match != null){
            if(localDate.isBefore(dateTime)){
                throw new IllegalArgumentException("This match has not taken place yet.");
            }
            match.setMatchScore(score1, score2);
        }
        //throwing an error if the match doesnt exist
        else if(match == null){
            throw new IllegalArgumentException("This match does not exist.");
        }
        //handling other errors
        else{
            throw new IllegalArgumentException("Error!");
        }
    }

    public List<Match> getUpcomingMatches(){
        //creating a list to store the upcoming matches
        List<Match> upcomingMatches = new ArrayList<>();

        //adding upcoming matches to the list
        for (Match m : listMatches){
            if(m.isUpcoming() == true){
                upcomingMatches.add(m);
            }
        }
        return upcomingMatches;
    }

    public List<Match> getMatchesOn(LocalDate dateTime){
        //creating a new list to store the matches on the specified date
        List<Match> matchesOn = new ArrayList<>();

        //adding specified matches to the list
        for(Match m : listMatches){
            if(m.getDateTime().equals(dateTime)){
                matchesOn.add(m);
            }
        }
        return matchesOn;
    }

    public List<Match> getMatchesFor(String teamName){
        //creating a new list to store matches for a specific team
        List<Match> matchesFor = new ArrayList<>();

        //adding matches for specific team to the list
        for(Match m : listMatches){
            if(m.getTeamA().getTeam().getName().equalsIgnoreCase(teamName)){
                matchesFor.add(m);
            }
            else if(m.getTeamB().getTeam().getName().equalsIgnoreCase(teamName)){
                matchesFor.add(m);
            }
        }
        return matchesFor;
    }

    public List<LineUp> getMatchLineUps(LocalDate dateTime){
        //creating a new list to store match lineups for a specific date
        List<LineUp> matchLineUps = new ArrayList<>();

        //adding the lineups for the specified date
        for (Match m : listMatches){
            if (m.getDateTime().equals(dateTime)){
                matchLineUps.add(m.getTeamA());
                matchLineUps.add(m.getTeamB());
            }
        }
        return matchLineUps;
    }


    public synchronized String toString(){
        StringBuilder output = new StringBuilder();
        output.append(String.format("Tournament name: %s\n", this.name));
        output.append(String.format("Start Date: %s\n", this.startDate));
        output.append(String.format("End Date: %s\n", this.endDate));

        return output.toString();
    }

}

File: PA2_Server/src/edu/ucdenver/Tournament/Tournament.py
import datetime
import json
from typing import List
from edu.ucdenver.Tournament.Country import Country
from edu.ucdenver.Tournament.Team import Team
from edu.ucdenver.Tournament.Player import Player
from edu.ucdenver.Tournament.Referee import Referee
from edu.ucdenver.Tournament.Lineup import Lineup
from edu.ucdenver.Tournament.Match import Match


class Tournament:
    """keeps track of a tournaments containing participating __participating_countries, __list_players,
    __list_teams, __list_referees, __list_matches, and lineups"""
    __name = ""
    __start_date = ""

    # --------------------------------------------------
    #                  constructor
    # --------------------------------------------------
    def __init__(self, name: str, start_date: datetime, end_date: datetime):
        """initializing tournament object"""
        self.__name = name
        self.__start_date = start_date
        self.__end_date = end_date
        self.__teamPlayers: List[Player] = []
        self.__list_teams: List[Team] = []
        self.__list_players: List[Player] = []
        self.__list_matches: List[Match] = []
        self.__list_line_ups: List[Lineup] = []
        self.__list_referees: List[Referee] = []
        self.countries: List[Country]= []
        self.__participating_countries: List[Country] = []
        self.__filename : "default"

    # --------------------------------------------------
    #               getters and setters
    # --------------------------------------------------
    def set_name(self, name: str):
        """setting tournament __name"""
        self.__name = name

    def set_start_date(self, __start_date: datetime):
        """setting start date"""
        self.__start_date = __start_date

    def set_end_date(self, __end_date: datetime):
        """setting end date"""
        self.__end_date = __end_date


    @property
    def filename(self):
        return self.__filename

    @filename.setter
    def filename(self, filename):
        self.__filename = filename


    # --------------------------------------------------
    #               class methods: admin
    # --------------------------------------------------

    def save_file(self, filename: str):

        temp = []
        for team in self.__list_teams:
            jsonStr = json.dumps(team.__dict__)
            temp.append(jsonStr)

        new_list = []
        for t in temp:
            final = {"Team Name": t[0], "Country": t[1], "Squad": t[3]}
            new_list.append(final)

        with open(self.__filename + '.json', 'w') as save_file:
            json.dump(new_list, save_file, indent=1)


        # var = self.__dict__
        #
        #
        # with open(self.__filename + '.json', 'w') as save_file:
        #     json.dump(var, save_file, indent=1)

    def load_file(self, filename: str):
        pass

    def add_country(self, country_name: str):
        """adding a participating country to the tournament"""
        country_exists = False

        # checking if country exists
        for country in self.countries:
            if country.name == country_name:
                country_exists = True

        # if country exists raise a value error
        if country_exists:
            raise ValueError("ERROR! This country already exists.")

        # if country does not exist create and add new country
        elif not country_exists:
            self.countries.append(Country(country_name))

    def add_team(self, team_name: str, country: str):
        """"adding a team to the tournament"""
        team_exists = False
        country_exists = False

        # check to see if team exists
        for team in self.__list_teams:
            if team.name == team_name:
                team_exists = True

        # check to see if country exists
        for c in self.__participating_countries:
            if c.name == country:
                country_exists = True

        # if team does not exist and country does exist add the team
        if team_exists is False and country_exists is True:
            self.__list_teams.append(Team(team_name, country))

        # if team does exist or country does not exist raise a value error
        elif team_exists is True or country_exists is False:
            if team_exists is True:
                raise ValueError("ERROR! This team already exists.")
            elif not country_exists:
                raise ValueError("ERROR! This country does not exist.")

    def add_referee(self, name: str, country: str):
        """adding a referee to the tournament"""
        referee_exists = False
        country_exists = False

        # checking to see if the referee already exists
        for referee in self.__list_referees:
            if referee.name == name:
                referee_exists = True

        # checking to see if the country exists
        for c in self.__participating_countries:
            if c.name == country:
                country_exists = True

        # if the country exists but the referee does not, add the referee
        if country_exists is True and referee_exists is False:
            self.__list_referees.append(Referee(name, country))

        # if the country does not exist or the referee already exists raise the appropriate error
        elif country_exists is False or referee_exists is True:
            if country_exists is False:
                raise ValueError("ERROR! This country does not exist.")
            elif referee_exists is True:
                raise ValueError("ERROR! This referee already exists.")

    def add_player(self, team_name: str, name: str, age: int, height: float, weight: float):
        """adding a player to a team in the tournament"""
        team_exists = False
        player_exists = False

        # checking to see if team exists and if player exists and adding player if team exists
        # and player does not already exist
        for team in self.__list_teams:
            if team.name == team_name:
                team_exists = True
                for player in team.squad:
                    player_exists = True
                    if player.name == name:
                        self.__list_players.append(team.add_player(name, age, height, weight))

        # raising the correct value error
        if team_exists is False or player_exists is True:
            if team_exists is False:
                raise ValueError("ERROR! This team does not exist.")
            elif player_exists is True:
                raise ValueError("ERROR! This player already exists.")

    def add_match(self, date_time: datetime, team_a: str, team_b: str):
        """adding a match involving two teams in the tournament"""
        team_a_exists = False
        team_b_exists = False

        # checking to see if team A and team B are equal
        # if they are equal, raise a value error
        if team_a.lower() == team_b.lower():
            raise ValueError("ERROR! Team A and Team B cannot be the same team.")

        # checking to see if team A and team B exist in the tournament
        # adding the match if they both exist
        for team1 in self.__list_teams:
            if team1.name == team_a:
                team_a_exists = True
                for team2 in self.__list_teams:
                    if team2.name == team_b:
                        team_b_exists = True
                        self.__list_matches.append(Match(date_time, team1, team2))

        # if one of the teams
        if team_a_exists is False or team_b_exists is False:
            raise ValueError("ERROR! One of these teams does not exist")

    def add_referee_to_match(self, date_time: datetime, referee_name: str):
        """adding a referee to a match occurring in the tournament"""
        referee_exists = False
        match_exists = False

        # checking to see if the match and referee exist before adding the
        # referee to the match
        for match in self.__list_matches:
            if match.date_time == date_time:
                match_exists = True
                for referee in self.__list_referees:
                    referee_exists = True
                    if referee.name.lower() is referee_name:
                        match.add_referee(referee)

        # raising the appropriate value error when referee or match doesn't exist
        if referee_exists is False or match_exists is False:
            if referee_exists is False:
                raise ValueError("ERROR! This referee does not exist.")
            elif match_exists is False:
                raise ValueError("ERROR! This match does not exist.")

    def add_player_to_match(self, date_time: datetime, team: str, player: str):
        """adding a player to a match lineup"""
        match_exists = False
        player_exists = False
        team_exists = False

        # checking if the match exists
        for match in self.__list_matches:
            if match.date_time == date_time:
                match_exists = True
                for t in self.__list_teams:
                    if t.name == team:
                        team_exists = True
                        # checking of player exists
                        for p in t.squad:
                            if p.name == player:
                                player_exists = True
                                match.add_player(p, t)

        if match_exists is False or player_exists is False or team_exists is False:
            if match_exists is False:
                raise ValueError("ERROR! This match does not exist.")
            elif team_exists is False:
                raise ValueError("ERROR! This team does not exist.")
            elif player_exists is False:
                raise ValueError("ERROR! This player does not exist")

    # Other functions
    def set_match_score(self, date_time: datetime, team_a_score: int, team_b_score: int):
        """setting the match score of a completed match"""
        match_exists = False

        for match in self.__list_matches:
            if match.date_time == date_time:
                match_exists = True
                match.set_score_team_a(team_a_score)
                match.set_score_team_b(team_b_score)
                match.set_match_score(team_a_score, team_b_score)

        # raising a value error and displaying message if match does not exist
        if match_exists is False:
            raise ValueError("ERROR! This match does not exist.")

    # --------------------------------------------------
    #            class methods: user/public
    # --------------------------------------------------

    def get_upcoming_matches(self):
        """returning matches that have not yet occurred in the tournament"""
        return self.__list_matches

    def get_matches_on(self, match_date: datetime):
        """returning matches that occur on a particular day of the tournament"""
        matches_on: List[Match] = []
        for match in self.__list_matches:
            m_date = match.date_time
            m_date = m_date.datetime.date()
            if m_date == match_date:
                matches_on.append(match)

        return matches_on

    def get_matches_for(self, team_name: str):
        """returning matches that involve a particular team participating in the tournament"""
        matches_for: List[Match] = []
        for match in self.__list_matches:
            team_a = match.team_a()
            team_b = match.team_b()
            if team_name.lower() == team_a.__name or team_name.lower() == team_b.__name:
                matches_for.append(match)
        return matches_for

    def get_match_line_ups(self, match_date: datetime):
        """returning the lineup for matches taking place on a specific day of the tournament"""
        match_lineups: List[Lineup] = []
        for match in self.__list_matches:
            if match.date_time == match_date:
                for line_up in self.__list_line_ups:
                    if match.team_a.__name == line_up.team.__name:
                        match_lineups.append(line_up)
                    if match.team_b.__name == line_up.team.__name:
                        match_lineups.append(line_up)

        return match_lineups

    # def add_lineup(self, team):
    # """adding a lineup to the tournament"""
    # self.__list_line_ups.append(Lineup(team))

    def __str__(self):
        """representing the tournament object as a string"""
        string = 'This is the {:s} tournament.|'.format(self.__name)
        return string
